src/librustc/ty/context.rs

```diff
@@ -1268,7 +1268,7 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
     }
 
     pub fn crates(self) -> Lrc<Vec<CrateNum>> {
-        self.all_crate_nums(LOCAL_CRATE)
+        Lrc::new(self.cstore.crates_untracked())
```

Comment:
I've changed the last commit so it also updates the return type of this as it is a new `Vec<..>` each time.

The reason that I changed this was because after speculatively loading a new crate for the purpose of diagnostics, the result of the `crates` query wouldn't change. This was fine for the purposes of adding a note like this:

```
    = note: the following trait defines an item `extern_baz`, perhaps you need to implement it:
            candidate #1: `baz::BazTrait`
```

But when adding a note like that below, it didn't work:

```
 help: the following trait is implemented but not in scope, perhaps add a `use` for it:
    |
 LL | use baz::BazTrait;
```

This is because the larger check that worked out whether this was appropriate to suggest called the `crates` query in a few places and the missing speculatively-loaded crates would cause the logic to always consider a trait from a speculatively-loaded trait unsuitable.

As I understand it, it isn't possible to invalidate the cache of the `crates` query after speculatively-loading crates (see [this message from @eddyb](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/subject/.2354618.20extern.20crates.20trait.20suggestions/near/136991244)).

The `crates` query just iterates over `self.metas` in the `cstore` and so will only ever contain the crates that exist when it was first invoked and then cached. I [wrote some more detail here](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/subject/.2354618.20extern.20crates.20trait.20suggestions/near/152109297) about my investigation into why only the first case got a suggestion, [@oli-obk then suggested stopping the `crates` function from being a query](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/subject/.2354618.20extern.20crates.20trait.20suggestions/near/152118827).